monitor.js

```diff
@@ -0,0 +1,94 @@
+const pull = require('pull-stream')
+
+function isForeignLeafFeedBranch(branch) {
+  if (branch.length !== 4) return false
+  if (branch[0].keys) return false // not interested in my branches
+  if (branch[0].purpose !== 'root') return false
+  if (branch[1].purpose !== 'v1') return false
+  if (branch[2].purpose.length !== 1) return false
+  return true
+}
+
+/**
+ * There may have been messages I had in my database but was unable to decrypt
+ * because I didn't know my root metafeed keys. Now that I know them, I should
+ * reindex my database and decrypt all messages.
+ *
+ * Then, we should restart monitoring because we have the root and now we can
+ * finally create DM keys for foreign leafs.
+ */
+function reindexAndRestart(ssb, encryptionFormat, drain) {
+  if (drain) drain.abort()
+  ssb.db.reindexEncrypted((err) => {
+    if (err) console.warn(err)
+
+    monitorForest(ssb, encryptionFormat)
+  })
+}
+
+/**
+ * Listen to a stream of all branches in the metafeed "forest" (collection of
+ * trees) and send some information such as keys to the box2 encryptionFormat.
+ *
+ * Box2 encryptionFormat uses keys from the metafeed tree to encrypt messages,
+ * this is why we need this information
+ */
+function monitorForest(ssb, encryptionFormat) {
+  encryptionFormat.getRootSigningKey((err, rootKeys) => {
+    if (err) console.warn(err)
+
+    let drain = null
+    pull(
+      ssb.metafeeds.branchStream({ old: true, live: true }),
+
+      // For all feeds that I own, add their keys to the box2 encryptionFormat
+      pull.asyncMap((branch, cb) => {
+        let foundMyRoot = false
+        for (let i = 0; i < branch.length; i++) {
+          const { id, keys, purpose } = branch[i]
+          if (keys) {
+            encryptionFormat.addSigningKeys(id, keys)
+            if (i === 0 && purpose === 'root' && !rootKeys) {
+              foundMyRoot = encryptionFormat.addSigningKeys('root', keys)
+              rootKeys = keys
+            }
+          }
+        }
+        if (foundMyRoot) reindexAndRestart(ssb, encryptionFormat, drain)
+        else cb(null, branch)
+      }),
+
+      // For all foreign leaf feeds, find their "mirrored" leaf feed in my tree,
+      // and build DM keys.
```

Comment:
Thinking out loud here, so lets say that we discover some we want to replicate. The replication scheduler will traverse the tree as will this module. So we basically have 2 concurrent things using the tree, one to EBT replicate and the other to add the keys. How do we make sure that either: the keys are added before we start decrypting the replicated messages, or that we reindex after we get a key?

Reply:
Really good point. I was a little bit aware of possible race conditions, including this one, but I wanted to just ship this PR.

We have an advantage here because the shard feed declares the existence of the leaf feed, and the shard feed is replicated before the leaf feed is replicated. So unless replication of the leaf feed is faster than the processing inside ssb-meta-feeds (i.e. processing the shard msg, building FeedDetails, and informing ssb-box2 via branchStream), then we won't have a problem. I don't think we will have a problem, but there's nothing *guaranteeing* we won't have a problem, and that's scary.

I accept suggestions how to fix this!

Reply:
Maybe we can leverage that keyring knows if a key is new or not. And if it's new we queue up a reindex. Probably with some kind of debouncing. There is additional thing there, if we use that kind of system what happens if we crash in the middle of a reindex. Maybe we should not store the key in keyring until we have reindexed? This system might also be different between group and DM keys. Mostly thinking out loud here.

Reply:
> Maybe we can leverage that keyring knows if a key is new or not.

Oh yeah great idea.

> There is additional thing there, if we use that kind of system what happens if we crash in the middle of a reindex.

There could be a feature in ssb-db2's reindexEncrypted where it is crash-resistant. Since reindexEncrypted takes no arguments, this is just a matter of storing a boolean in some file (or just the file existence can be the "true" boolean) and then once ssb-db2 starts up it resumes reindexEncrypted if that file exists. Note also that reindexEncrypted does not need to be atomic. Any subset of encrypted records that got decrypted is some partial progress, and then after app restart, the progress can continue.

Also, another thing I notice is that it seems reindexEncrypted is not very efficient, at least not with jitdb. If there are encrypted records at seq 6, 14, 18, 22, and assume max seq is 100 then it'll reindex jitdb from seq 6–100, then reindex 14–100, then 18–100, then 22–100. It could just pick the minimum (6) and do that once.

Reply:
wierd, I have to recheck the db2 logic, but you are right, we should make that crash-resistant. Just have to be careful we don't end up in a infinite loop because of a bug somewhere :)

Reply:
Lets create an issue in db2 for this. The solution here looks good.

Reply:
Created db2 issue number 407